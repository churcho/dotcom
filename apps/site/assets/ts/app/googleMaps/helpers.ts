// @ts-ignore
import stopMarker from "../../../static/images/icon-map-stop-marker.svg";
// @ts-ignore
import stationMarker from "../../../static/images/icon-map-station-marker.svg";
// @ts-ignore
import selectedStopMarker from "../../../static/images/icon-map-stop-marker-hover.svg";
// @ts-ignore
import selectedStationMarker from "../../../static/images/icon-map-station-marker-hover.svg";
// @ts-ignore
import currentLocationMarker from "../../../static/images/icon-current-location-marker.svg";
import { iconSize } from "../../../js/google-map/helpers";
import mapStyles from "../../../js/google-map/styles";
import { MarkerData, MapData } from "./__googleMaps";

export const stopIcon = (selected: boolean): string =>
  selected ? selectedStopMarker : stopMarker;
export const stationIcon = (selected: boolean): string =>
  selected ? selectedStationMarker : stationMarker;

export const buildIconFromSVG = (
  svg: string,
  size: string
): google.maps.Icon => {
  const sizeInt = iconSize(size);

  const encoded = window.btoa(svg);

  return {
    url: `data:image/svg+xml;base64, ${encoded}`,
    scaledSize: new window.google.maps.Size(sizeInt, sizeInt),
    origin: new window.google.maps.Point(0, 0),
    labelOrigin: new window.google.maps.Point(sizeInt / 2, sizeInt / 2)
  };
};

export const buildMarkerIcon = (
  data: MarkerData,
  selected: boolean
): google.maps.Icon | undefined => {
  if (data.icon) {
    if (data.icon!.includes("station")) {
      return buildIconFromSVG(stationIcon(selected), data.size);
    }
    if (data.icon!.includes("stop")) {
      return buildIconFromSVG(stopIcon(selected), data.size);
    }
    if (data.icon!.includes("current-location")) {
      return buildIconFromSVG(currentLocationMarker, data.size);
    }
  }
  return undefined;
};

export const setMapDefaults = (
  map: google.maps.Map,
  initialData: MapData
): void => {
  const options = {
    center: new window.google.maps.LatLng(
      initialData.default_center.latitude,
      initialData.default_center.longitude
    ),
    ...initialData.dynamic_options
  };
  const styles = mapStyles as google.maps.MapTypeStyle[];
  // Hide stop icons on the transit layer
  styles.push({
    featureType: "transit",
    elementType: "labels.icon",
    stylers: [{ visibility: "off" }]
  });
  map.setOptions({ styles, ...options });
  map.setZoom(initialData.zoom || 17);

  if (initialData.layers && initialData.layers.transit) {
    const layer = new window.google.maps.TransitLayer();
    layer!.setMap(map);
  }
};
