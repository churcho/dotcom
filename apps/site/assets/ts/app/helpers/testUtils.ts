// eslint-disable-next-line import/prefer-default-export
export const createReactRoot = (): void => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';
};
