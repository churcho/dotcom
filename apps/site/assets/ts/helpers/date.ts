// Returns a date in the form of "August 7, 2019"
export const formattedDate = (unformatted: string): string => {
  const [year, month, day] = unformatted
    .split(/-/)
    .map(part => Number.parseInt(part, 10));

  // Remember that months in JS are 0-indexed for some reason, hence "month - 1".
  const parsedDate = new Date(year, month - 1, day);
  return parsedDate.toLocaleDateString("en-US", {
    year: "numeric",
    month: "long",
    day: "numeric",
    timeZone: "UTC"
  });
};

// Returns a time in the form of "05:16 AM"
export const shortTime = (date: string): string =>
  new Date(date).toLocaleTimeString("en-US", {
    hour: "2-digit",
    minute: "2-digit"
  });

export default formattedDate;
