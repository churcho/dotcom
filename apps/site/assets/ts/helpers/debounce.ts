// Debounce is meant to be used with event listeners. Use it to delay a call
// until after a burst of events, such as keyboard input.
//
//   usage:
//
//     const callback = (ev: Event) => {}
//     document.addEventListener("keyup", debounce(callback, 250))
//

const debounce = (callback: Function, delay: number): EventListener => {
  let timer: number | undefined;

  return (...args) => {
    window.clearTimeout(timer);
    timer = window.setTimeout(() => callback(...args), delay);
  };
};

export default debounce;
