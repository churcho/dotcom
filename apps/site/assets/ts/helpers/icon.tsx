import renderSvg from "./render-svg";
// @ts-ignore
import accessibleIconSvg from "../../static/images/icon-accessible-default.svg";
// @ts-ignore
import blueLineIconSvg from "../../static/images/icon-blue-line-default.svg";
// @ts-ignore
import busIconSvg from "../../static/images/icon-mode-bus-default.svg";
// @ts-ignore
import commuterRailIconSvg from "../../static/images/icon-mode-commuter-rail-default.svg";
// @ts-ignore
import greenLineIconSvg from "../../static/images/icon-green-line-default.svg";
// @ts-ignore
import greenBLineIconSvg from "../../static/images/icon-green-line-b-default.svg";
// @ts-ignore
import greenCLineIconSvg from "../../static/images/icon-green-line-c-default.svg";
// @ts-ignore
import greenDLineIconSvg from "../../static/images/icon-green-line-d-default.svg";
// @ts-ignore
import greenELineIconSvg from "../../static/images/icon-green-line-e-default.svg";
// @ts-ignore
import mattapanLineIconSvg from "../../static/images/icon-mattapan-line-default.svg";
// @ts-ignore
import orangeLineIconSvg from "../../static/images/icon-orange-line-default.svg";
// @ts-ignore
import parkingIconSvg from "../../static/images/icon-parking-default.svg";
// @ts-ignore
import redLineIconSvg from "../../static/images/icon-red-line-default.svg";

export const accessibleIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, accessibleIconSvg);

export const blueLineIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, blueLineIconSvg);

export const busIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, busIconSvg);

export const commuterRailIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, commuterRailIconSvg);

export const greenLineIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, greenLineIconSvg);

export const greenBLineIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, greenBLineIconSvg);

export const greenCLineIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, greenCLineIconSvg);

export const greenDLineIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, greenDLineIconSvg);

export const greenELineIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, greenELineIconSvg);

export const mattapanLineIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, mattapanLineIconSvg);

export const orangeLineIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, orangeLineIconSvg);

export const parkingIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, parkingIconSvg);

export const redLineIcon = (className: string = ""): JSX.Element =>
  renderSvg(className, redLineIconSvg);
