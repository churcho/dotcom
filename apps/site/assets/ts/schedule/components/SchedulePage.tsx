import React, { ReactElement } from "react";
import { SchedulePageData } from "./__schedule";
import PDFSchedules from "./PDFSchedules";
import ContentTeasers from "./ContentTeasers";
import HoursOfOperation from "./HoursOfOperation";

interface Props {
  schedulePageData: SchedulePageData;
}

const SchedulePage = ({
  schedulePageData: { pdfs, teasers, hours }
}: Props): ReactElement<HTMLElement> => (
  <>
    <ContentTeasers teasers={teasers} />
    <PDFSchedules pdfs={pdfs} />
    <HoursOfOperation hours={hours} />
  </>
);
export default SchedulePage;
