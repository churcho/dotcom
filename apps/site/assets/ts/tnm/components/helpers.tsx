import React, { KeyboardEvent } from "react";

export const handleEnterKeyPress = (
  e: KeyboardEvent,
  onClick: Function
): void => {
  if (e.key === "Enter") {
    onClick();
  }
};

export const renderSvg = (className: string, svgText: string): JSX.Element => (
  <span
    className={className}
    // eslint-disable-next-line react/no-danger
    dangerouslySetInnerHTML={{ __html: svgText }}
  />
);
