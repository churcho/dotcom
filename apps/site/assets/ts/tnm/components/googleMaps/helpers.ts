// @ts-ignore
import stopMarker from "../../../../static/images/icon-map-stop-marker.svg";
// @ts-ignore
import stationMarker from "../../../../static/images/icon-map-station-marker.svg";
// @ts-ignore
import selectedStopMarker from "../../../../static/images/icon-map-stop-marker-hover.svg";
// @ts-ignore
import selectedStationMarker from "../../../../static/images/icon-map-station-marker-hover.svg";
// @ts-ignore
import currentLocationMarker from "../../../../static/images/icon-current-location-marker.svg";
import { iconSize } from "../../../../js/google-map/helpers";

import { MarkerData } from "../__tnm";

interface GoogleMapMarkerSVG {
  url: string;
  scaledSize: google.maps.Size;
  origin: google.maps.Point;
  anchor: google.maps.Point;
  labelOrigin: google.maps.Point;
}

export const stopIcon = (selected: boolean): string =>
  selected ? selectedStopMarker : stopMarker;
export const stationIcon = (selected: boolean): string =>
  selected ? selectedStationMarker : stationMarker;

export const buildIconFromSVG = (
  svg: string,
  size: string
): GoogleMapMarkerSVG => {
  const sizeInt = iconSize(size);

  const encoded = window.btoa(svg);

  return {
    url: `data:image/svg+xml;base64, ${encoded}`,
    scaledSize: new window.google.maps.Size(sizeInt, sizeInt),
    origin: new window.google.maps.Point(0, 0),
    anchor: new window.google.maps.Point(sizeInt / 2, sizeInt / 2),
    labelOrigin: new window.google.maps.Point(sizeInt / 2, sizeInt / 2)
  };
};

export const buildMarkerIcon = (
  data: MarkerData,
  selected: boolean
): GoogleMapMarkerSVG | undefined => {
  if (data.icon) {
    if (data.icon!.includes("station")) {
      return buildIconFromSVG(stationIcon(selected), data.size);
    }
    if (data.icon!.includes("stop")) {
      return buildIconFromSVG(stopIcon(selected), data.size);
    }
    if (data.icon!.includes("current-location")) {
      return buildIconFromSVG(currentLocationMarker, data.size);
    }
  }
  return undefined;
};
