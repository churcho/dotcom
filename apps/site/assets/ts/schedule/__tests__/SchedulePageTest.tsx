import React from "react";
import renderer from "react-test-renderer";
import { createReactRoot } from "../../app/helpers/testUtils";
import SchedulePage from "../components/SchedulePage";

const pdfs = [
  {
    url: "https://mbta.com/example-pdf.pdf",
    title: "Route 1 schedule PDF"
  }
];

const teasers = `<div><a href="http://some-link">Some teaser from CMS></a></div>`;

const hours = `<div class="m-schedule-page__sidebar-hours">  <h3 class="hours-period-heading">Monday to Friday</h3>
<p class="hours-directions">
  <span class="hours-direction-name">Inbound</span>
  <span class="hours-time">04:17A-12:46A</span>
</p>
<p class="hours-directions">
  <span class="hours-direction-name">Outbound</span>
  <span class="hours-time">05:36A-01:08A</span>
</p>
</div>`;

it("it renders", () => {
  createReactRoot();
  const tree = renderer
    .create(<SchedulePage schedulePageData={{ pdfs, teasers, hours }} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
