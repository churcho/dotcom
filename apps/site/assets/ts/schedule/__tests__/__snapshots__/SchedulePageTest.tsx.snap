// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
Array [
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<div><a href=\\"http://some-link\\">Some teaser from CMS></a></div>",
      }
    }
  />,
  <h3
    aria-controls="panel-pdfs"
    aria-expanded={true}
    className="c-expandable-block__header"
    id="header-pdfs"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    PDF Schedules
    <span
      className="c-expandable-block__header-caret"
      dangerouslySetInnerHTML={
        Object {
          "__html": "&#xF106;",
        }
      }
    />
  </h3>,
  <div
    aria-labelledby="header-pdfs"
    className="c-expandable-block__panel"
    id="panel-pdfs"
    role="region"
    tabIndex={0}
  >
    <a
      className="m-schedule-sidebar__pdfs"
      href="https://mbta.com/example-pdf.pdf"
      rel="noopener noreferrer"
      target="_blank"
    >
      <span
        aria-hidden="true"
        className="notranslate c-svg__icon-pdf"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Route 1 schedule PDF
    </a>
  </div>,
  <noscript>
    <style>
      #header-pdfs { display: none; }
    </style>
    <h3
      className="c-expandable-block__header"
    >
      PDF Schedules
      <span
        className="c-expandable-block__header-caret"
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#xF106;",
          }
        }
      />
    </h3>
    <div
      className="c-expandable-block__panel"
      role="region"
    >
      <a
        className="m-schedule-sidebar__pdfs"
        href="https://mbta.com/example-pdf.pdf"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          aria-hidden="true"
          className="notranslate c-svg__icon-pdf"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
        Route 1 schedule PDF
      </a>
    </div>
  </noscript>,
  <h3
    aria-controls="panel-hours"
    aria-expanded={false}
    className="c-expandable-block__header"
    id="header-hours"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    Hours of Operation
    <span
      className="c-expandable-block__header-caret"
      dangerouslySetInnerHTML={
        Object {
          "__html": "&#xF107;",
        }
      }
    />
  </h3>,
  <noscript>
    <style>
      #header-hours { display: none; }
    </style>
    <h3
      className="c-expandable-block__header"
    >
      Hours of Operation
      <span
        className="c-expandable-block__header-caret"
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#xF106;",
          }
        }
      />
    </h3>
    <div
      className="c-expandable-block__panel"
      role="region"
    >
      <div
        className="m-schedule-page__sidebar-hours"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<div class=\\"m-schedule-page__sidebar-hours\\">  <h3 class=\\"hours-period-heading\\">Monday to Friday</h3>
<p class=\\"hours-directions\\">
  <span class=\\"hours-direction-name\\">Inbound</span>
  <span class=\\"hours-time\\">04:17A-12:46A</span>
</p>
<p class=\\"hours-directions\\">
  <span class=\\"hours-direction-name\\">Outbound</span>
  <span class=\\"hours-time\\">05:36A-01:08A</span>
</p>
</div>",
          }
        }
      />
    </div>
  </noscript>,
]
`;
